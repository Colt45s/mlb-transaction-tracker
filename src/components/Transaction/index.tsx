import React, { useRef, useCallback, useMemo } from 'react';
import { pickColorByCodeType } from '../../utils/color';
import { Row } from '../transactionTracker';
import { PlayerIcon } from '../PlayerIcon';
import { TransactionDetail } from '../TransactionDetail';

type Props = {
  transaction: Row;
  className?: string;
};

export function Transaction(props: Props) {
  const imgSrc = useMemo(
    () =>
      `https://content.mlb.com/images/headshots/current/60x60/${props.transaction.player_id}@2x.png`,
    [props.transaction.player_id]
  );

  const imgRef = useRef<HTMLImageElement>(null);

  const handleImgError = useCallback(() => {
    if (imgRef) {
      if (imgRef.current) {
        imgRef.current.src =
          'https://content.mlb.com/images/headshots/current/60x60/generic@2x.png';
      }
    }
  }, []);

  const tagColor = useMemo(
    () => pickColorByCodeType(props.transaction.type_cd) || '#343a40',
    [props.transaction.type_cd]
  );

  return (
    <>
      <PlayerIcon
        imgRef={imgRef}
        imgSrc={imgSrc}
        handleImgError={handleImgError}
      />
      <TransactionDetail transaction={props.transaction} tagColor={tagColor} />
    </>
  );
}
