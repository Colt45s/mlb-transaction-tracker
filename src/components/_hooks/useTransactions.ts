import { useCallback, useRef, Dispatch } from 'react';
import { getStartDate, getCurrentDate, getYesterday } from '../../utils/date';
import { Action, Row } from '../transactionTracker';

const ENDPOINT =
  'https://lookup-service-prod.mlb.com/json/named.transaction_all.bam';

export function useTransactions(dispatch: Dispatch<Action>) {
  const startDateRef = useRef('');

  const fetchTransactions = useCallback(async () => {
    const endDate = startDateRef.current
      ? getYesterday(startDateRef.current)
      : getCurrentDate();
    const startDate = getStartDate(endDate);
    const url = `${ENDPOINT}?sport_code='mlb'&start_date=${startDate}&end_date=${endDate}`;

    startDateRef.current = startDate;
    const response = await fetch(url);

    if (!response.ok) {
      throw new Error(`Request failed: ${response.status}`);
    }

    const data = await response.json();
    const { queryResults } = data.transaction_all;

    if (!parseInt(queryResults.totalSize)) {
      // fetch prev week
      await fetchTransactions();
    } else {
      const row: Row | Row[] = queryResults.row;
      let processRow: Row[];

      if (row instanceof Array) {
        processRow = row;
      } else {
        processRow = [row];
      }

      dispatch({
        type: 'SUCCESS',
        payload: {
          items: processRow
            .slice()
            .reverse()
            .filter((i: any) => i.orig_asset_type === 'PL')
        }
      });
    }
  }, [dispatch]);

  return useCallback(async () => {
    try {
      await fetchTransactions();
    } catch (e) {
      dispatch({ type: 'ERROR', payload: { error: e } });
    }
  }, [dispatch, fetchTransactions]);
}
