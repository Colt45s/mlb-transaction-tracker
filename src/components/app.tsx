import React, { Suspense } from 'react';
import { createGlobalStyle } from 'styled-components';
import reset from 'styled-reset';
import 'react-loader-spinner/dist/loader/css/react-spinner-loader.css';

const TransactionTracker = React.lazy(() => import('./transactionTracker'));

const GlobalStyle = createGlobalStyle`
  ${reset}
  @import url('https://fonts.googleapis.com/css?family=Baloo+Paaji+2&display=swap');
  html {
    font-size: 62.5%;
  }
  body {
    line-height: 1.5;
    font-size: 1.4rem;
    font-family: 'Baloo Paaji 2', cursive;
    background-color: #f8f8f8;
  }
`;

export function App() {
  return (
    <>
      <GlobalStyle />
      <Suspense fallback={null}>
        <TransactionTracker />
      </Suspense>
    </>
  );
}
