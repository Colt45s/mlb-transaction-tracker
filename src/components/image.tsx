import React from 'react';

type Props = {
  imgRef: React.RefObject<HTMLImageElement>;
  src: string;
  alt?: string;
  style?: React.CSSProperties;
  handleImgError: () => void;
};

export const Image = React.memo((props: Props) => (
  <img
    ref={props.imgRef}
    style={props.style}
    src={props.src}
    onError={props.handleImgError}
  />
));
