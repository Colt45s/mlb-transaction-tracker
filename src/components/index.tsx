import React from 'react'
import { createGlobalStyle } from 'styled-components'
import reset from 'styled-reset'
import InfiniteLoader from './infiniteLoader'

const GlobalStyle = createGlobalStyle`
  ${reset}
  body {
    line-height: 1.75;
    font-size: 1em;
    font-family: -apple-system, BlinkMacSystemFont, "Hiragino Kaku Gothic ProN", Meiryo, sans-serif;
  }
`

export default () => (
  <>
    <GlobalStyle />
    <InfiniteLoader />
  </>
)
