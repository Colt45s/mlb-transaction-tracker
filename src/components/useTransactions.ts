import { useState, useCallback } from 'react';
import { getStartDate, getCurrentDate, getYesterday } from '../utils/date';
import { Dispatch } from 'redux';
import { useDispatch } from 'react-redux';
import { useSelector } from '../utils/useSelector';
import { Action } from '../actions/transaction';
import { fetchTransactionSuccess } from '../actions/transaction';
import { Row } from '../reducers/transaction';

const ENDPOINT =
  'https://lookup-service-prod.mlb.com/json/named.transaction_all.bam';

export function useTransactions(): [Row[], () => Promise<void>, boolean, any] {
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState<any>(null);
  const [baseDate, setBaseDate] = useState('');

  const items = useSelector(state => state.transaction.items);
  const dispatch = useDispatch<Dispatch<Action>>();

  const fetchTransactions = useCallback(
    async bd => {
      const endDate = bd ? getYesterday(bd) : getCurrentDate();
      const startDate = getStartDate(endDate);
      const url = `${ENDPOINT}?sport_code='mlb'&start_date=${startDate}&end_date=${endDate}`;

      setLoading(true);

      try {
        const response = await fetch(url);
        const data = await response.json();
        const { queryResults } = data.transaction_all;

        setLoading(false);

        if (!parseInt(queryResults.totalSize)) {
          // fetch prev week
          fetchTransactions(startDate);
        } else {
          dispatch(
            fetchTransactionSuccess(
              queryResults.row
                .slice()
                .reverse()
                .filter((i: any) => i.orig_asset_type === 'PL')
            )
          );
        }
        setBaseDate(startDate);
      } catch (e) {
        setLoading(false);
        setError(e.messsage);
      }
    },
    [dispatch]
  );

  const fetcher = useCallback(async () => {
    fetchTransactions(baseDate);
  }, [baseDate, fetchTransactions]);

  return [items, fetcher, loading, error];
}
