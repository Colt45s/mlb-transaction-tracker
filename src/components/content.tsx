import React from 'react'
import { Row } from './useTransactions'
import Transaction from './transaction'

type Props = {
  transaction: Row
  isLoaded: boolean
  error: any
}

export default React.memo((props: Props) => {
  if (props.error) {
    return <div>Error !</div>
  }

  return props.isLoaded ? (
    <Transaction transaction={props.transaction} />
  ) : (
    <div>Loading</div>
  )
})
