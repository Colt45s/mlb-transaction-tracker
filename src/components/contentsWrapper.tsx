import React, { CSSProperties } from 'react'
import { Row } from './useTransactions'
import Content from './content'
import styled from 'styled-components'

type Props = {
  style: CSSProperties
  transaction: Row
  isLoaded: boolean
  error: any
  className?: string
}

const View = React.memo((props: Props) => (
  <div style={props.style} className={props.className}>
    <div>
      <Content
        transaction={props.transaction}
        isLoaded={props.isLoaded}
        error={props.error}
      />
    </div>
  </div>
))

export default styled(View)`
  border-bottom: 1px solid #eee;

  > * {
    padding: 0.5em;
  }
`
