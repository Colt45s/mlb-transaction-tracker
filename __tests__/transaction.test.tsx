import React from 'react'
import { render, cleanup } from 'react-testing-library'
import 'jest-dom/extend-expect'

import Transaction from '../src/components/transaction'
import { Row } from '../src/components/useTransactions'

afterEach(cleanup)

test('snapshot props', () => {
  const props: { transaction: Row } = {
    transaction: {
      player: 'test',
      team: 'hou',
      note: 'test',
      from_team: 'astros',
      type: '',
      type_cd: 'RET'
    }
  }

  const { asFragment } = render(<Transaction {...props} />)

  expect(asFragment()).toMatchSnapshot()
})
