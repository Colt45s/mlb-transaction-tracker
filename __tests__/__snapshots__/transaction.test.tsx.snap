// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot props 1`] = `
<DocumentFragment>
  <div
    class="sc-bwzfXH ideAXj"
  >
    <div
      class="contents"
    >
      <h3>
        test
      </h3>
      <span
        class="sc-bdVaJa ldkgbU"
      />
      <p>
        From : astros
      </p>
      <p>
        To : hou
      </p>
      <p>
        test
      </p>
    </div>
  </div>
</DocumentFragment>
`;
